File: v2/examples/js/urihandler-v2.js
import { createHash } from 'node:crypto';

export function parseUri(uri) {
  const match = String(uri).match(/^(?<scheme>[a-z][a-z0-9+.-]*):\/\/(?<target>[^/?#]+)(?<path>\/[^?#]*)?(?:\?(?<query>[^#]*))?(?:#(?<fragment>.*))?$/i);
  if (!match) throw new Error(`Invalid URI: ${uri}`);
  const segments = (match.groups.path || '/').split('/').filter(Boolean).map(decodeURIComponent);
  return {
    package: match.groups.scheme,
    target: decodeURIComponent(match.groups.target),
    segments,
    query: Object.fromEntries(new URLSearchParams(match.groups.query || '')),
    fragment: match.groups.fragment || null,
    raw: uri,
  };
}

export function normalizeUri(d) {
  return `${d.package}://${d.target}/${d.segments.map(encodeURIComponent).join('/')}`;
}

export function translate(d) {
  const [resource, operation, ...rest] = d.segments;
  return {
    route: [d.package, resource, operation],
    args: [d.target, ...rest],
    package: d.package,
    target: d.target,
    resource,
    operation,
    descriptor: { ...d, normalized: normalizeUri(d) },
  };
}

export function validate(t, registry) {
  return !!registry?.[t.package]?.[t.resource]?.[t.operation];
}

export function hashUri(normalized) {
  return createHash('sha256').update(normalized).digest('hex');
}

export function resolve(t, registry, runtimeCache = new Map()) {
  const key = hashUri(t.descriptor.normalized);
  if (runtimeCache.has(key)) return runtimeCache.get(key);
  const fn = registry?.[t.package]?.[t.resource]?.[t.operation];
  if (typeof fn !== 'function') throw new Error(`Unresolved route: ${t.route.join('.')}`);
  runtimeCache.set(key, fn);
  return fn;
}

export async function dispatch(uri, registry, payload, runtimeCache = new Map()) {
  const descriptor = parseUri(uri);
  const translation = translate(descriptor);
  if (!validate(translation, registry)) throw new Error('Route validation failed');
  const fn = resolve(translation, registry, runtimeCache);
  return await fn(translation.target, translation.args.slice(1), payload, translation.descriptor);
}

File: v2/examples/python/urihandler_v2.py
import hashlib
import re
from urllib.parse import parse_qsl, unquote, quote

URI_RE = re.compile(r'^(?P<scheme>[a-z][a-z0-9+.-]*)://(?P<target>[^/?#]+)(?P<path>/[^?#]*)?(?:\?(?P<query>[^#]*))?(?:#(?P<fragment>.*))?$', re.I)

def parse_uri(uri: str):
    m = URI_RE.match(str(uri))
    if not m:
        raise ValueError(f'Invalid URI: {uri}')
    segments = [unquote(s) for s in (m.group('path') or '/').split('/') if s]
    return {
        'package': m.group('scheme'),
        'target': unquote(m.group('target')),
        'segments': segments,
        'query': dict(parse_qsl(m.group('query') or '')),
        'fragment': m.group('fragment') or None,
        'raw': uri,
    }

def normalize_uri(d: dict):
    return f"{d['package']}://{d['target']}/{'/'.join(quote(s, safe='') for s in d['segments'])}"

def translate(d: dict):
    resource, operation, *rest = d['segments']
    return {
        'route': [d['package'], resource, operation],
        'args': [d['target'], *rest],
        'package': d['package'],
        'target': d['target'],
        'resource': resource,
        'operation': operation,
        'descriptor': {**d, 'normalized': normalize_uri(d)},
    }

def validate(t: dict, registry: dict):
    return callable(registry.get(t['package'], {}).get(t['resource'], {}).get(t['operation']))

def hash_uri(normalized: str):
    return hashlib.sha256(normalized.encode('utf-8')).hexdigest()

def resolve(t: dict, registry: dict, runtime_cache: dict | None = None):
    runtime_cache = runtime_cache if runtime_cache is not None else {}
    key = hash_uri(t['descriptor']['normalized'])
    if key in runtime_cache:
        return runtime_cache[key]
    fn = registry.get(t['package'], {}).get(t['resource'], {}).get(t['operation'])
    if not callable(fn):
        raise KeyError(f"Unresolved route: {'.'.join(t['route'])}")
    runtime_cache[key] = fn
    return fn

def dispatch(uri: str, registry: dict, payload=None, runtime_cache: dict | None = None):
    descriptor = parse_uri(uri)
    translation = translate(descriptor)
    if not validate(translation, registry):
        raise ValueError('Route validation failed')
    fn = resolve(translation, registry, runtime_cache)
    return fn(translation['target'], translation['args'][1:], payload, translation['descriptor'])
